src/hooks/useWrapCallback.ts

```diff
@@ -17,58 +19,117 @@ export enum WrapType {
   UNWRAP,
 }
 
+export enum WrapState {
+  UNKNOWN,
+  PENDING,
+  WRAPPED,
+  UNWRAPPED,
+}
+
 const NOT_APPLICABLE = { wrapType: WrapType.NOT_APPLICABLE }
+interface UseWrapCallbackReturn {
+  wrapType: WrapType
+  execute?: () => Promise<void>
+  inputError?: string
+  wrapState?: WrapState
+  setWrapState?: (wrapState: WrapState) => void
+}
+
 /**
  * Given the selected input and output currency, return a wrap callback
  * @param inputCurrency the selected input currency
  * @param outputCurrency the selected output currency
  * @param typedValue the user input value
  */
-export default function useWrapCallback(
-  inputCurrency: Currency | undefined,
-  outputCurrency: Currency | undefined,
-  typedValue: string | undefined
-): { wrapType: WrapType; execute?: undefined | (() => Promise<void>); inputError?: string } {
+export function useWrapCallback(
+  currencies: { INPUT?: Currency | undefined; OUTPUT?: Currency | undefined },
+  isGnosisTrade = false,
+  typedValue?: string // can be also obtained from the SwapState
+): UseWrapCallbackReturn {
   const { chainId, account } = useActiveWeb3React()
   const nativeCurrency = useNativeCurrency()
   const nativeCurrencyWrapperToken = useWrappingToken(nativeCurrency)
   const nativeCurrencyWrapperContract = useNativeCurrencyWrapperContract()
-  const balance = useCurrencyBalance(account ?? undefined, inputCurrency)
+
+  const balance = useCurrencyBalance(account ?? undefined, currencies.INPUT as Currency)
   const { t } = useTranslation()
+  const [wrapState, setWrapState] = useState(WrapState.UNKNOWN)
   // we can always parse the amount typed as the input currency, since wrapping is 1:1
-  const inputAmount = useMemo(() => tryParseAmount(typedValue, inputCurrency, chainId), [
-    inputCurrency,
-    typedValue,
-    chainId,
-  ])
+  const inputAmount = useMemo(() => {
+    const isOutputCurrencyNative = !!(currencies.OUTPUT && Currency.isNative(currencies.OUTPUT))
+    const inputCurrency =
+      isGnosisTrade && isOutputCurrencyNative
+        ? wrappedCurrency(currencies.OUTPUT as Currency, chainId as ChainId)
+        : currencies.INPUT
+    return tryParseAmount(typedValue, inputCurrency, chainId)
+  }, [currencies, typedValue, isGnosisTrade, chainId])
+
+  const outputAmount = useMemo(() => {
+    const isInputCurrencyNative = !!(currencies.INPUT && Currency.isNative(currencies?.INPUT))
+    const outputCurrency =
+      isGnosisTrade && isInputCurrencyNative
+        ? wrappedCurrency(currencies.INPUT as Currency, chainId as ChainId)
+        : currencies.OUTPUT
+    return tryParseAmount(typedValue, outputCurrency, chainId)
+  }, [currencies, typedValue, isGnosisTrade, chainId])
```

Comment:
This entire section can be combined into a single `useMemo` and then destructued. They both depend on the same variables.